polyfill/lib/ecmascript.mjs

```diff
@@ -566,6 +566,20 @@ export const ES = ObjectAssign({}, ES2019, {
     }
     return ES.Call(from, TemporalTimeZone, [temporalTimeZoneLike]);
   },
+  ToTemporalTimeZone: (temporalTimeZoneLike) => {
+    if (typeof temporalTimeZoneLike === 'object' && temporalTimeZoneLike) {
+      return temporalTimeZoneLike;
+    }
```

Comment:
The key question is what we want to happen if `Temporal.TimeZone.from` is overridden. I'm working based on the idea that the overridden function is only called with strings, not with objects (custom or built-in). I feel it would be confusing for `absolute.toDateTime({ ... custom time zone object ... })` to potentially *not* use its argument for the time zone calculations. Let me know if you want to discuss further.

Reply:
That's a coherent idea, but a different one from what I've been assuming, which is that the monkeypatched `Temporal.TimeZone.from()` still does have to handle objects correctly. There's a bit of discussion about this in #294.

Reply:
I read through #294 and the notes from the referenced meetings, but see nothing that would lean that way. On the contrary, some of the comments (e.g. "I realized that the monkeypatcher has to get the handling of objects just right, and when to convert the argument to a string, in order to match the original Temporal.TimeZone.from().") seem to lean "my way".

I do think the approach proposed here leads to more predictable behaviour, but happy to go whichever way the champions decide.

Reply:
Maybe I should back up a moment and realize that I don't understand why `absolute.toDateTime({ ... custom object ... })` would not use its argument for time zone calculations. If I understood that, maybe I'd understand your objection better :smile: 

The reason I'm commenting is because it seems to me that ToTemporalTimeZone does exactly the same thing as TimeZoneFrom: ToTemporalTimeZone returns the object if it's an object, and call TimeZoneFrom with the argument converted to a string otherwise. TimeZoneFrom calls Temporal.TimeZone.from(). Temporal.TimeZone.from() returns the object if it's an object, and constructs a new instance from the argument converted to a string otherwise.

Reply:
> TimeZoneFrom calls Temporal.TimeZone.from(). Temporal.TimeZone.from() returns …

Except that you don't know what Temporal.TimeZone.from() returns. That's exactly the point of this whole exercise; we need to consider what happens when author-overridden methods do weird or confusing things. For example,

```js
const original = Temporal.TimeZone.from;
Temporal.TimeZone.from = function(arg) {
  if (arg == "UTC") return original(arg);
  return new CustomTimeZone(String(arg)); // stringification could be implicit somewhere as well
};
```

doesn't look all that suspicious, but it does break custom objects. It's my experience that authors will come up with much stranger things than you could ever imagine :)

I guess the main question is if there's a good reason for author-overridden `Temporal.TimeZone.from` to do anything to object arguments besides returning them straight back to the caller. If not (and I think this is the case), passing them in is only an opportunity for the overridden method to introduce bugs, and (less importantly) it's somewhat less performant because of the extra call into author JS.

Reply:
OK, I understand this line of reasoning now. I think the conclusion that we had previously reached about author-overridden `Temporal.TimeZone.from` behaving badly was "well, don't do that then". I agree it would reduce the opportunities for bugs...

On the other hand, it doesn't save author-overridden `Temporal.TimeZone.from` the trouble of implementing that logic, either. They still have to return objects straight back to the caller. And I think if it is the case that they would ever not want to do that, this behaviour would be surprising.

I think it should be fine to merge this and change it back if it leads to problems in the future.